conseil-common/src/main/scala/tech/cryptonomic/conseil/common/rpc/RpcClient.scala

```diff
@@ -0,0 +1,176 @@
+  *   import org.http4s.circe.CirceEntityEncoder._
+  *
+  *   implicit val contextShift = IO.contextShift(ExecutionContext.global)
+  *
+  *   // Define case class for JSON-RPC method and response
+  *   case class ApiMethodParams(id: Int)
+  *   case class ApiMethodResponse(name: String)
+  *
+  *   // Create stream with the request
+  *   val request = Stream(
+  *     RpcRequest(
+  *       jsonrpc = "2.0",
+  *       mathod = "apiMethod",
+  *       params = ApiMethodParams(id = 1)
+  *       id = "id_of_the_request"
+  *     )
+  *   )
+  *
+  *   // Create JSON-RPC client
+  *   val client = new RpcClient[IO]("endpoint", 10, httpClient)
+  *
+  *   // Call api method with
+  *   val response: Stream[IO, ApiMethodResponse] =
+  *     client.stream[ApiMethodParams, ApiMethodResponse](batchSize = 10)(request)
+  * }}}
+  *
+  * @param endpoint JSON-RPC server url
+  * @param maxConcurrent Max concurrent processes used to fetch data from JSON-RPC server
+  * @param httpClient Http4s http client
+  * @param headers Additional http headers, e.g. authorization
+  */
+class RpcClient[F[_]: Concurrent](
+    endpoint: String,
+    maxConcurrent: Int,
+    httpClient: Client[F],
+    headers: Header*
+) extends Http4sClientDsl[F]
+    with LazyLogging {
+
+  /**
+    * Method that converts [[fs2.Stream]] of JSON-RPC requests into [[fs2.Stream]]] of desired result objects,
+    * or throws the exception with the error. Request [[fs2.Stream]]] can only contain calls to one API method
+    * and can be batched to speed up the response time.
+    *
+    * @param batchSize The size of the batched request in single HTTP call
+    * @param request [[fs2.Stream]]] of the [[RpcRequest]]
+    *
+    * @return [[fs2.Stream]]] that contains result from JSON-RPC server call.
+    */
+  def stream[P, R](batchSize: Int)(request: Stream[F, RpcRequest[P]])(
+      implicit decode: EntityDecoder[F, Seq[RpcResponse[R]]],
+      encode: EntityEncoder[F, List[RpcRequest[P]]]
+  ): Stream[F, R] =
+    request
+      .chunkN(batchSize)
+      .evalTap(requests => debug(s"Call RPC in a batch of: ${requests.size}"))
+      // From JSON-RPC documentation: The Server MAY process a batch rpc call as a set of concurrent tasks, 
+      //                              processing them in any order and with any width of parallelism.
+      // So we also can process the responses in unordered order.
+      .mapAsyncUnordered(maxConcurrent) { chunks =>
+        httpClient
+          .expect[Seq[RpcResponse[R]]](
+            Method.POST(chunks.toList, Uri.unsafeFromString(endpoint), headers: _*)
+          )
+      }
+      .flatMap(Stream.emits)
+      .flatMap {
+        case RpcResponse(_, _, Some(error), _) =>
+          Stream.raiseError[F](error)
+        case RpcResponse(_, Some(result), _, _) =>
+          Stream(result)
+        case response =>
+          Stream.raiseError[F](
+            new UnsupportedOperationException(
```

Comment:
Is this "unsupported operation" or "unexpected response", because for me these are two different things and we are putting them together.

Reply:
According to the JSON-RPC spec, this operation is impossible. So which exception should I use in your opinion?

Reply:
Well... if this operation is not supported, then change the body message to `Unsupported operation: $x`. I don't see how unsupported operation leads to unexpected response?